vm/os-windows.cpp

```diff
@@ -304,6 +304,61 @@ void handle_ctrl_c() {
   SetConsoleCtrlHandler(factor::ctrl_handler, TRUE);
 }
 
+static volatile bool stop_on_ctrl_break = false;
+static HANDLE ctrl_break_thread = nullptr;
+
+static DWORD WINAPI ctrl_break_thread_proc(LPVOID mainThread) {
+  bool ctrl_break_handled = false;
+  while (true) {
+    if (factor::stop_on_ctrl_break) {
+      if (GetAsyncKeyState(VK_CANCEL) >= 0) { /* Ctrl-Break is released. */
+        ctrl_break_handled = false;  /* Wait for the next press. */
+      } else if (!ctrl_break_handled) {
+        /* This proc runs in its own thread without stopping the main
+           thread. Since in practice nobody uses the multi-VM stuff yet,
+           we just interrupt the first VM we can get. */
+        FACTOR_ASSERT(thread_vms.size() > 0);
+        THREADHANDLE thd = thread_vms.begin()->first;
+        factor_vm* vm = thread_vms.begin()->second;
+        /* Check if the VM thread has the same Id as the thread Id of the
+           currently active window. Note that thread Id is not a handle. */
+        DWORD fgThreadId = GetWindowThreadProcessId(GetForegroundWindow(),
+                                                    nullptr);
+        if (!atomic::load(&vm->safepoint.fep_p)
+            && (fgThreadId == vm->thread_id)
+        ) {
+          vm->safepoint.enqueue_fep(vm, true);
+          wake_up_thread(thd); /* Try to wake up the thread. */
+          ctrl_break_handled = true;
+        }
+      }
+    }
+    Sleep(10);
+  }
+  return 0;
+}
+
+void ignore_ctrl_break() {
+  stop_on_ctrl_break = false;
+}
+
+void handle_ctrl_break() {
+  stop_on_ctrl_break = true;
+  if (ctrl_break_thread == nullptr) {
+    DisableProcessWindowsGhosting();
+    ctrl_break_thread = CreateThread(nullptr, 0, factor::ctrl_break_thread_proc,
+                                     nullptr, 0, nullptr);
+  }
+}
+
+void factor_vm::primitive_enable_ctrl_break() {
+  handle_ctrl_break();
+}
+
+void factor_vm::primitive_disable_ctrl_break() {
+  ignore_ctrl_break();
+}
+
```

Comment:
You can just put the contents of the functions ignore/handle_ctrl_break in primitive_enable/disable_ctrl_break. No need to add extra wrappings.